### PR TITLE
maint(ci): add mypy type checking to Travis

### PR DESCRIPTION
**Pull request by oscarbenjamin**
#### References to other Issues or PRs

Previous work is in #20140 and #18244

#### Brief description of what is fixed or changed

Adds a mypy job on Travis so that mypy is tested on all PRs.

#### Other comments

I deliberately didn't do this at the time of #18244 because I found mypy somewhat flakey. I think it can help improve the codebase but it is not necessarily that intuitive. If we enable this then all contributors will need to deal with mypy (it is not as easy as flake8 so help will be needed).

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by Maelstrom6** [comment]
I was just starting to do this after installing mypy later today. Apologies for the "non-staticness" of `stochastic_process_types.py`. 

I'm getting lots of errors of the form `Name 'simplify' already defined (by an import)`. Was this detected in Travis?

**Comment by asmeurer** [comment]
How strict is mypy about needing to know the types of everything? I haven't used it much myself, but my experience from watching colleagues use it is that you often run into a situation where you can't figure out how to type something to make it happy. Have we run into that situation with SymPy?

I also find it annoying if it requires changes like the one made in test_unify.py. I don't like it when linters force you to rewrite your code in some less readable way in order to make them happy.

File: .travis.yml [hunk 1]
```diff
@@ -31,6 +31,10 @@ jobs:
       dist: xenial
       env:
         - TEST_FLAKE8="true"
+    - python: 3.8
+      dist: xenial
+      env:
```
Comment (asmeurer) [inline .travis.yml:36]: Any reason not to merge this with the previous build? Travis limits the number of concurrent jobs, so if two jobs can be merged into one that is preferable.

Comment (oscarbenjamin) [reply .travis.yml:36]: Using separate jobs reduces the total time taken for this stage of the build. Running jobs concurrently is generally a good thing although it does include a small overhead for job startup.

**Comment by oscarbenjamin** [comment]
> I'm getting lots of errors of the form `Name 'simplify' already defined (by an import)`. Was this detected in Travis?

I've seen those errors before and I believe I have already added some `type:ignore` hints for them. I think it's a bug in mypy but it's also IMO a design mistake in the sympy codebase that we end up with something like:
```
from sympy.simplify.simplify import simplify
from sympy.simplify import simplify
from sympy import simplify
```
All three of those imports give the simplify function even though the 3rd and 2nd could also refer to the simplify package or its simplify submodule. This has also caused problems with e.g. sphinx.

**Comment by oscarbenjamin** [comment]
> How strict is mypy about needing to know the types of everything?

In general it is not necessary to type most things. For example almost all of the sympy codebase is untyped and mypy is fine with that.

To get the real benefit of mypy though you do need to add types. Once you start adding types mypy will do more checks but it can do them using inference to some extent so it will only force you to add types in cases where the inference fails. Since these are the only cases I really fixed in #18244 you can see what those changes were. Mainly the inference fails on something like an empty dict.

There are some patterns in sympy that mypy doesn't like such as classes where the instance attributes are different for each instance e.g.:
https://github.com/sympy/sympy/blob/153864a6ed7432256022a13c78499e558d04fa31/sympy/vector/coordsysrect.py#L235-L237

**Comment by asmeurer** [comment]
So mypy isn't going to complain that something isn't typed, just not give useful warnings for it?

**Comment by oscarbenjamin** [comment]
There are some situations where mypy complains that something isn't typed. The main example that came up is something like
https://github.com/sympy/sympy/blob/153864a6ed7432256022a13c78499e558d04fa31/sympy/core/sympify.py#L28
Type inference fails for a completely empty dict like this and mypy will complain.

Mostly though the idea is "gradual typing" which is supposed to mean that you don't have to add any types but mypy only fully checks things that are typed:
https://mypy.readthedocs.io/en/stable/existing_code.html

**Comment by asmeurer** [comment]
Regarding the `a, b, c = 'abc'` thing, apparently mypy disallows it on purpose https://github.com/python/mypy/issues/6406. Frankly, that sort of thing makes me uneasy about using mypy. They apparently don't care about having rules that have blatant false positives (compared to something like pyflakes, which tries very hard to avoid them).

**Comment by asmeurer** [comment]
I guess the question is, what happens when mypy needs a type for something, and it happens to be the sort of thing that isn't possible to type currently? Can you just ignore it with an Any type, or are you forced to refactor the code to make it more typable?

**Comment by oscarbenjamin** [comment]
It is possible to silence mypy with `type:ignore` in all cases that I have seen. That might be annoying but I think it will help when reviewing code. If a PR adds `type:ignore` then as a reviewer you can see that in the diff and can ask about it.

Using mypy does mean deciding to disallow some ways of writing the code though (that is its purpose). The idea should be that you use it if you think that it enforcing those limitations on the codebase will lead to a better result on balance.

Personally I was not interested in mypy or typing when they were introduced because I didn't see how they would improve my own code. Similarly I had never used flake8 or any of the related tools. That has changed now because I maintain a 750kloc codebase built on disparate contributions that vary enormously in quality. I now see these tools as ways of automating some part of the review process which is actually where the bottleneck usually lies for a large open source project.

I also now see how in a codebase the size of sympy's taking advantage of Python's dynamic typing very often just leads to brokenness. Before mypy/typing came along the Python mantra was all about duck-typing. I think duck-typing is great but it seems few people understand how it is supposed to work. Duck-typing works great when you have a well defined protocol such as "iterable" or "sequence" etc.

You have to actually define what that protocol/interface is supposed to be though or otherwise duck-typing just means that your code only half works and will quickly break in any of the situations that isn't explicitly tested. SymPy's main classes like Expr can not be duck-typed: they have far too many methods, attributes and properties. If a function is written for Expr then we should not allow any non-Expr to be passed in.

There are lots of places in sympy where types are mixed up in incoherent ways. Fixing those once they have become entrenched is not easy. If we can prevent that sort of thing in future then that's great but it doesn't solve the mess of matrix classes with non-Expr elements or the incoherent mix of Expr and Poly in PolyMatrix or the fact that many sympy classes have the wrong superclasses etc.

On the other hand we don't really get the benefit of mypy until we add type-hints everywhere. The reason I didn't immediately pursue adding mypy to Travis after #18244 and putting type hints around the codebase was that I realised that type hints are not really inherited by subclasses. I had hoped that simply adding the hints to the relevant methods on Basic, Expr, Set etc would mean that those same methods would be checked for all of their subclasses. That's not how it works though: you have to add the type hints to every method in every subclass or otherwise mypy won't check them.

**Comment by asmeurer** [comment]
> On the other hand we don't really get the benefit of mypy until we add type-hints everywhere. The reason I didn't immediately pursue adding mypy to Travis after #18244 and putting type hints around the codebase was that I realised that type hints are not really inherited by subclasses. I had hoped that simply adding the hints to the relevant methods on Basic, Expr, Set etc would mean that those same methods would be checked for all of their subclasses. That's not how it works though: you have to add the type hints to every method in every subclass or otherwise mypy won't check them.

I think mypy supports some sort of stub file that defines type signatures. It may be possible to automatically generate stub files for subclasses based on the type signatures of the superclasses.

File: sympy/stats/stochastic_process_types.py [hunk 2]
```diff
@@ -349,35 +351,35 @@ def _sanity_checks(cls, state_space, trans_probs):
...
             # Range object doesn't want to give a symbolic size
             # so we do it ourselves.
-            if isinstance(state_space, Range):
-                ss_size = ceiling((state_space.stop - state_space.start) / state_space.step)
+            if isinstance(v_state_space, Range):
+                ss_size = ceiling((v_state_space.stop - v_state_space.start) / v_state_space.step)
             else:
-                ss_size = len(state_space)
+                ss_size = len(v_state_space)
             if ss_size != trans_probs.shape[0]:
```
Comment (czgdp1807) [inline sympy/stats/stochastic_process_types.py:378]:
```suggestion
            if ss_size != v_trans_probs.shape[0]:
```

File: sympy/stats/stochastic_process_types.py [hunk 3]
```diff
@@ -349,35 +351,35 @@ def _sanity_checks(cls, state_space, trans_probs):
...
+            v_trans_probs = _matrix_checks(trans_probs)
+            v_state_space = _state_converter(Range(trans_probs.shape[0]))
 
         elif trans_probs is None:
-            state_space = _state_converter(state_space)
+            v_state_space = _state_converter(state_space)
             if isinstance(state_space, Range):
                 _n = ceiling((state_space.stop - state_space.start) / state_space.step)
             else:
                 _n = len(state_space)
```
Comment (czgdp1807) [inline sympy/stats/stochastic_process_types.py:366]:
```suggestion
                _n = Integer(len(v_state_space))
```

File: sympy/stats/stochastic_process_types.py [hunk 4]
```diff
@@ -349,35 +351,35 @@ def _sanity_checks(cls, state_space, trans_probs):
...
-            trans_probs = _matrix_checks(trans_probs)
-            state_space = _state_converter(Range(trans_probs.shape[0]))
+            v_trans_probs = _matrix_checks(trans_probs)
+            v_state_space = _state_converter(Range(trans_probs.shape[0]))
 
         elif trans_probs is None:
-            state_space = _state_converter(state_space)
+            v_state_space = _state_converter(state_space)
             if isinstance(state_space, Range):
                 _n = ceiling((state_space.stop - state_space.start) / state_space.step)
```
Comment (czgdp1807) [inline sympy/stats/stochastic_process_types.py:364]:
```suggestion
            if isinstance(v_state_space, Range):
                _n = ceiling((v_state_space.stop - v_state_space.start) / v_state_space.step)
```

File: sympy/stats/stochastic_process_types.py [hunk 5]
```diff
@@ -349,35 +351,35 @@ def _sanity_checks(cls, state_space, trans_probs):
...
             # Range object doesn't want to give a symbolic size
             # so we do it ourselves.
-            if isinstance(state_space, Range):
-                ss_size = ceiling((state_space.stop - state_space.start) / state_space.step)
+            if isinstance(v_state_space, Range):
+                ss_size = ceiling((v_state_space.stop - v_state_space.start) / v_state_space.step)
             else:
-                ss_size = len(state_space)
-            if ss_size != trans_probs.shape[0]:
+                ss_size = len(v_state_space)
```
Comment (czgdp1807) [inline sympy/stats/stochastic_process_types.py:377]:
```suggestion
                ss_size = Integer(len(v_state_space))
```

**Comment by czgdp1807** [comment]
I have resolved the conflicts. Please see if there are any more changes to be made. ping @oscarbenjamin

**Pull request closed without merging**